tests/v4l2decode.cpp

```diff
@@ -330,6 +329,153 @@ static bool displayOneVideoFrameAndroid(int32_t fd, int32_t index)
+)
+{
+    struct display * d = (struct display * )data;
+
+    if (strcmp(interface, "wl_compositor") == 0)
+        d->compositor.reset((struct wl_compositor *)wl_registry_bind(registry, id,
+                                                   &wl_compositor_interface, 1), wl_compositor_destroy);
+    else if (strcmp(interface, "wl_shell") == 0)
+        d->shell.reset((struct wl_shell *)wl_registry_bind(registry, id,
+	                                               &wl_shell_interface, 1), wl_shell_destroy);
+}
+
+void frameRedrawCallback(void *data,
+	                                       struct wl_callback *callback, uint32_t time)
+{
+    *(bool *)data = false;
+    wl_callback_destroy(callback);
+}
+
+bool ensureWindow(unsigned int width, unsigned int height)
+{
+    struct display * const d = &waylandDisplay;
+
+    if (!d->surface) {
+        d->surface.reset(wl_compositor_create_surface(d->compositor.get()), wl_surface_destroy);
+        if (!d->surface)
+            return false;
+    }
+
+    if (!d->shell_surface) {
+        d->shell_surface.reset(wl_shell_get_shell_surface(d->shell.get(), d->surface.get()),
+                                                                       wl_shell_surface_destroy);
+        if (!d->shell_surface)
+            return false;
+        wl_shell_surface_set_toplevel(d->shell_surface.get());
+    }
+    return true;
+}
+
+bool createWaylandDisplay()
+{
+    SharedPtr<VADisplay> display;
+    struct display *d = &waylandDisplay;
+    struct wl_registry_listener registry_listener = {
+        registryHandle,
+        NULL,
+    };
+
+    d->display.reset(wl_display_connect(NULL), wl_display_disconnect);
+    if (!d->display) {
+        return false;
+    }
+    wl_display_set_user_data(d->display.get(), d);
+    struct wl_registry *registry = wl_display_get_registry(d->display.get());
+    wl_registry_add_listener(registry, &registry_listener, d);
+    wl_display_dispatch(d->display.get());
+    return true;
+}
+
+bool vaPutSurfaceWayland(struct wl_buffer *buffer,
+                                              const VARectangle *srcRect,
+                                              const VARectangle *dstRect)
+{
+    struct wl_callback *callback;
+    struct display * const d = &waylandDisplay;
+    struct wl_callback_listener frame_callback_listener = {frameRedrawCallback};
+
+    if (redrawPending) {
+        wl_display_flush(d->display.get());
+        while (redrawPending) {
+            wl_display_dispatch(d->display.get());
+        }
+    }
+
+    if (!ensureWindow(dstRect->width, dstRect->height))
+        return false;
+
+    wl_surface_attach(d->surface.get(), buffer, 0, 0);
+    wl_surface_damage(d->surface.get(), dstRect->x,
+		dstRect->y, dstRect->width, dstRect->height);
+    wl_display_flush(d->display.get());
+    redrawPending = true;
+    callback = wl_surface_frame(d->surface.get());
+    wl_callback_add_listener(callback, &frame_callback_listener,&redrawPending);
```

Comment:
frame_callback_listener is local variable, i'm not sure whether it is ok to set it to a callback listener.